test/datalog_tests/simple3.dl

```diff
@@ -0,0 +1,208 @@
+    var res = vec_with_capacity(list.length());
+    into_nonnull_vec_inner(list, res);
+    res
+}
+
+function into_nonnull_vec_inner(list: NonNullList<'N>, vec: mut Vec<'N>) {
+    vec.push(list.node);
+    match (list.nxt) {
+        NULL -> (),
+        NonNullNxt{nxt} -> into_nonnull_vec_inner(nxt.deref(), vec)
+    }
+}
+
+// A helper function for the temporary workaroud
+function nonnull_list_nxt(list: NonNullList<'N>): NonNullNxt<'N> {
+	list.nxt
+}
+
+function nonnull_list_nth(list: NonNullList<'N>, i: usize): Option<'N> {
+	if (i == 0) {
+		Some{ list.node }
+	} else {
+		match (list.nxt) {
+			NonNullNxt { nxt_nonnull_list_ref } -> {
+				var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
+				nonnull_list_nth(nxt_nonnull_list, i-1)
+			},
+			NULL -> { None }
+		}
+	}
+}
+
+function nonnull_list_append(list: NonNullList<'N>, item: 'N): NonNullList<'N> {
+	match (list.nxt) {
+		NonNullNxt { nxt_nonnull_list_ref } -> {
+			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
+			NonNullList { list.node, NonNullNxt { ref_new(nonnull_list_append(nxt_nonnull_list, item)) } }
+		},
+		NULL -> { 
+			// Create a NonNullList with two items
+			var nxt = NonNullNxt { ref_new(NonNullList { item, NULL }) };
+			NonNullList { list.node, nxt }
+		}
+	}
+}
+
+function nonnull_list_reverse(list: NonNullList<'N>): NonNullList<'N> {
+	match (list.nxt) {
+		NonNullNxt { nxt_nonnull_list_ref } -> {
+			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
+			var reversed_nxt_nonnull_list = nonnull_list_reverse(nxt_nonnull_list);
+			nonnull_list_append(reversed_nxt_nonnull_list, list.node)
+		},
+		NULL -> { list }
+	}
+}
+
+function nonnull_list_map(list: NonNullList<'N>, f: function('N): 'M): NonNullList<'M> {
+	var new_node = f(list.node);
+	var mapped_nonnull_nxt = match (list.nxt) {
+		NonNullNxt {nxt_nonnull_list_ref} -> {
+			var nxt_nonnull_list = deref(nxt_nonnull_list_ref);
+			var mapped_nonnull_list = nonnull_list_map(nxt_nonnull_list, f);
+			NonNullNxt { ref_new(mapped_nonnull_list) }
+		},
+		NULL -> NULL
+	};
+	NonNullList { 
+		new_node, 
+		mapped_nonnull_nxt
+	}
+}
+
+output relation NonNullListTest(description: string)
+
+function test_non_null_list(): NonNullList<usize> {
+    NonNullList{ 1, NonNullNxt{ ref_new(NonNullList{ 2, NonNullNxt{ ref_new(NonNullList{ 3, NULL }) } }) } }
+}
+
+NonNullListTest("from_nonnull_vec([1,2,3,4,5]): ${from_nonnull_vec([1,2,3,4,5]).to_string_debug()}").
+NonNullListTest("test_non_null_list(): ${test_non_null_list().to_string_debug()}").
+NonNullListTest("into_nonnull_vec(test_non_null_list()): ${into_nonnull_vec(test_non_null_list()).to_string_debug()}").
+NonNullListTest("nonnull_list_append(test_non_null_list(), 4): ${nonnull_list_append(test_non_null_list(), 4).to_string_debug()}").
+NonNullListTest("nonnull_list_map(test_non_null_list(), +1): ${nonnull_list_map(test_non_null_list(), |x| x + 1).to_string_debug()}").
+
+// `List` in Formula representation:
```

Comment:
is this stuff tested anywhere?
Or is it here just to trigger the generics bug?

Reply:
Nope, the functionality is not tested. It's mostly to test type inference.  Ideally, there would be functional tests too, but I don't want to spend more time on this atm.

@qishen, if you ever finish this library and have some tests for it, please share them with us!

Reply:
Thanks for the comments and suggestions! I just found some issues with my current implementation of the linked list that returns the wrong result. I'll share it with you later once I fix it.